numba_dpex/core/types/usm_ndarray_type.py

```diff
@@ -31,46 +31,53 @@ def __init__(
         aligned=True,
         addrspace=address_space.GLOBAL,
     ):
+        # Creating SyclDevice from filter_string is expensive. So, USMNdArray should be able to
```

Comment:
As itself the change to check for `SyclDevice` is fine. Right now since `dpctl.SyclDevice` is not a supported type inside dpjit, the only code path that can use `SyclDevice` is when a `USMNdArray` is created from the `typeof` function.

Reply:
This code path happens every time you call numba jitted function and is the main source of calling overhead.

Reply:
I am specifically talking about passing a `SyclDevice` object to the `UsmNdArray` constructor. For example, refer to https://github.com/IntelPython/numba-dpex/blob/dd110998669127305f2d10dbd78114b8b2c13d7d/numba_dpex/dpnp_iface/arrayobj.py#L208 and other overloads where we create a `UsmNdArray` or `DpnpNdArray`. As a `SyclDevice` object can never be passed in to the overload right now, when we create the `UsmNdArray` we will not right now be passing in a `SyclDevice`.

Reply:
I'm not sure how `overload` is related to the issue.
The issue itself is about overhead of calling numba-dpex functions.
It was found that overhead is caused by two calls: `filter_string` and `SyclDevice()`. `filter_string` overhead was fixed in `dpctl` by caching. On `SyclDevice()` we have agreed that we would not create new device every time. Instead we would take already existing from `usm_ndarray` (https://github.com/IntelPython/numba-dpex/issues/945#issuecomment-1453559034).

That's what this fix exactly do. Takes already created `SyclDevice` from `usm_ndarray`.
Do you see other ways to avoid `SyclDevice` re-creation?

Reply:
Ok. I think I got it. Instead of passing `device` and `queue` from `typeof` we could pass only `queue` and then extract device from it. Is that what you are suggesting?

Reply:
@diptorupd I've update PR, but coverage fails. Should I update `dpctl` version in `meta.yaml` to `14.3dev0`?

Reply:
@AlexanderKalistratov It will not solve internal CI failure.

@oleksandr-pavlyk when is the next tag for dpctl due? We need a new tag on TC servers to get the `dpctl.get_device_cached` function.